sdk/storage/azure-storage-common-data-movement/src/main/java/com/azure/storage/common/data/movement/PathScanner.java

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.common.data.movement;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.*;
+import java.util.stream.BaseStream;
+import java.util.stream.Stream;
+
+public class PathScanner {
+    private final String basePath;
+    private final boolean isDir;
+
+    public PathScanner(String path, boolean isDir) {
+        basePath = path;
+        this.isDir = isDir;
+    }
+
+    public Flux<String> scan(boolean skipSubdirectories) {
+        // Iterate through the path if it's a directory
+        if (isDir) {
+            // Set up Mono for recursing through directories
+            Mono<Path> basePath = Mono.just(Paths.get(this.basePath));
+
+            // Return a Flux constructed from expanding path
+            return basePath.expand(path ->
+                Flux.using(() -> Files.list(path),
+                    Flux::fromStream,
+                    BaseStream::close)
+                    .onErrorResume((e) -> {
+                        // Ignore errors resulting from trying to enumerate contents of
+                        // a file as if it were a directory, but log other errors
+                        if (!(e instanceof NotDirectoryException))
+                            System.err.println(e);
+
+                        // If set to skip subdirectories, continue processing; else,
+                        // pass an error which will stop the Flux stream
+                        if (skipSubdirectories)
+                            return Mono.empty();
+                        else
+                            return Mono.error(e);
+                    }))
+                // Remove the folders from the final returned values
+                .filter(path -> !Files.isDirectory(path))
```

Comment:
Why are we removing the folders from the returned values? Do we do this in .NET as well?

Reply:
Yeah, in .NET folders are also removed since the folders aren't passed to the upload functions, just the files are. (Is that correct?)

Reply:
I think we should change both to return files and directories, when we have directories sometimes we may want to put a directory there explicitly (for example for file share or file datalake accounts)

Reply:
I'm not sure how far is that but users importing to HNS may want to preserve ACLs, dir can have different ACLs than files inside.

Reply:
Gotcha, I'll change this. Assuming I should mod this in .NET as well?

Reply:
Yes let's mod this in .NET as well